text/0182-otlp-remote-parent.md

```diff
@@ -0,0 +1,108 @@
+# Export SpanContext.IsRemote in OTLP
+
+Update OTLP to indicate whether a span's parent is remote.
+
+## Motivation
+
+It is sometimes useful to post-process or visualise only entry-point spans: spans which either have no parent (trace roots), or which have a remote parent.
+For example, the Elastic APM solution highlights entry-point spans (Elastic APM refers to these as "transactions") and surfaces these as top-level operations
+in its user interface.
+
+The goal is to identify the spans which represent a request that is entering a service, or originating within a service, without having to first assemble the
+complete distributed trace as a DAG (Directed Acyclic Graph). It is trivially possible to identify trace roots, but it is not possible to identify spans with
+remote parents.
+
+Here is a contrived example distributed trace, with a border added to the entry-point spans:
+
+```mermaid
+graph TD
+    subgraph comments_service
+    POST_comments(POST /comment)
+    POST_comments --> comments_send(comments send)
+    end
+
+    subgraph auth_service
+    POST_comments --> POST_auth(POST /auth)
+    POST_auth --> LDAP
+    end
+
+    subgraph user_details_service
+    POST_comments --> GET_user_details(GET /user_details)
+    GET_user_details --> SELECT_users(SELECT FROM users)
+    end
+
+    subgraph comments_inserter
+    comments_send --> comments_receive(comments receive)
+    comments_receive --> comments_process(comments process)
+    comments_process --> INSERT_comments(INSERT INTO comments)
+    end
+
+    style POST_comments stroke-width:4
+    style POST_auth stroke-width:4
+    style GET_user_details stroke-width:4
+    style comments_receive stroke-width:4
+```
+
+## Explanation
+
+The OTLP encoding for spans has a boolean `parent_span_is_remote` field for identifying whether a span's parent is remote or not.
+All OpenTelemetry SDKs populate this field, and backends may use it to identify a span as being an entry-point span.
+A span can be considered an entry-point span if it has no parent (`parent_span_id` is empty), or if `parent_span_is_remote` is true.
+
+## Internal details
+
+The first part would be to update the trace protobuf, adding a `boolean parent_span_is_remote` field to the
+[`Span` message](https://github.com/open-telemetry/opentelemetry-proto/blob/b43e9b18b76abf3ee040164b55b9c355217151f3/opentelemetry/proto/trace/v1/trace.proto#L84).
+
+[`SpanContext.IsRemote`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote) identifies whether span context has been propagated from a remote parent.
+The OTLP exporter in each SDK would need to be updated to record this in the new `parent_span_is_remote` field.
```

Comment:
I don't think this would be sufficient. `SpanContext.IsRemote` can only be true for the _parent span context_, passed to StartSpan. Once the new span is started, the parent span context may be lost (reduced to `parent_span_id` field), so the Exporter won't have access to it.

Reply:
> Once the new span is started, the parent span context may be lost (reduced to parent_span_id field), so the Exporter won't have access to it.

Are you saying that `SpanContext.IsRemote` can be wrong? Do you have an example in mind where only the parent span ID is retained in the parent span context? The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote) doesn't suggest anything like that.

If that's the case, then I suppose implementers would have to record the parent span context's `SpanContext.IsRemote` property on the child span at creation time.

Reply:
Parent span context is usually used to create the current span's context, by copying flags and trace id, and then to populate the `parent_span_id` field (and NOT copying IsRemote flag). After that the parent context is not needed for the current span, so by the time the span is finished and sent to exporters the parent's IsRemote indicator is lost unless it is captured explicitly in the span. In either case, `span.context.IsRemote` is always going to be false, by definition.

Reply:
> After that the parent context is not needed for the current span, so by the time the span is finished and sent to exporters the parent's IsRemote indicator is lost unless it is captured explicitly in the span.

Depending on how you read the spec for ReadableSpan, the full parent SpanContext must be captured, and most (all?) SDKs do it, or capture at least a `has_remote_parent` boolean (not is_remote but has_remote_parent!). E.g. see https://github.com/dotnet/runtime/issues/65660.

<https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#additional-span-interfaces>:

> Readable span: A function receiving this as argument MUST be able to access all information that was added to the span, as listed [in the API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span).

and at the linked API spec:

> Spans encapsulate:
> ...
> * A parent span in the form of a [Span](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span), [SpanContext](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spancontext), or null

Reply:
So it may be implied, but the spec does not explicitly require an API like hasRemoteParent, are you suggesting to add it?

Reply:
Yes, I think that would be a good idea.

Reply:
> After that the parent context is not needed for the current span, so by the time the span is finished and sent to exporters the parent's IsRemote indicator is lost unless it is captured explicitly in the span.

Doesn't that conflict with the wording that @Oberon00 referenced above? i.e. `MUST be able to access all information that was added to the span` - which to me means that the parent SpanContext (as originally specified, fully intact) must be available on a span when it reaches the exporter.

> In either case, span.context.IsRemote is always going to be false, by definition.

Sorry, I think I was unclear my description. What I meant was that the span's *parent* `SpanContext.IsRemote()` would be used by the exporter. It the parent SpanContext is required to be accessible in a ReadableSpan, then we can use its IsRemote property.

Reply:
If you interpret the spec strictly (which you should normally do), the full Context should be available which includes the whole parent Span (not readable). While that would totally work, I don't think any SDK has implemented it that way (.NET has the full parent ReadWriteSpan if there was one provided though, but .NET is divergent in other ways). So I think it would make sense to clarify this point in a non-breaking way, so that most current SDK implementation can be considered conformant without requiring changes. Unless someone has a strong use case for wanting the full parent Context available.

Reply:
I had forgotten about this. Now created https://github.com/open-telemetry/opentelemetry-specification/pull/3257